extensions/ql-vscode/src/queries-panel/query-discovery.ts

```diff
@@ -1,154 +1,116 @@
+    private readonly queryPackDiscovery: QueryPackDiscoverer,
   ) {
-    super("Query Discovery", extLogger);
+    super("Query Discovery", `**/*${QUERY_FILE_EXTENSION}`);
 
-    this.onDidChangeQueriesEmitter = this.push(new EventEmitter<void>());
-    this.push(workspace.onDidChangeWorkspaceFolders(this.refresh.bind(this)));
-    this.push(this.watcher.onDidChange(this.refresh.bind(this)));
-  }
-
-  public get queries(): Array<FileTreeDirectory<string>> | undefined {
-    return this.results?.queries;
+    this.push(
+      this.queryPackDiscovery.onDidChangeQueryPacks(
+        this.recomputeAllQueryLanguages.bind(this),
+      ),
+    );
   }
 
   /**
-   * Event to be fired when the set of discovered queries may have changed.
+   * Event that fires when the set of queries in the workspace changes.
    */
   public get onDidChangeQueries(): Event<void> {
-    return this.onDidChangeQueriesEmitter.event;
-  }
-
-  protected async discover(): Promise<QueryDiscoveryResults> {
-    const workspaceFolders = getOnDiskWorkspaceFoldersObjects();
-    if (workspaceFolders.length === 0) {
-      return {
-        queries: [],
-        watchPaths: [],
-      };
-    }
-
-    const queries = await this.discoverQueries(workspaceFolders);
-
-    return {
-      queries,
-      watchPaths: workspaceFolders.map((f) => f.uri),
-    };
-  }
-
-  protected update(results: QueryDiscoveryResults): void {
-    this.results = results;
-
-    this.watcher.clear();
-    for (const watchPath of results.watchPaths) {
-      // Watch for changes to any `.ql` file
-      this.watcher.addWatch(new RelativePattern(watchPath, "**/*.{ql}"));
-      // need to explicitly watch for changes to directories themselves.
-      this.watcher.addWatch(new RelativePattern(watchPath, "**/"));
-    }
-    this.onDidChangeQueriesEmitter.fire();
+    return this.onDidChangePathsEmitter.event;
   }
 
   /**
-   * Discover all queries in the specified directory and its subdirectories.
-   * @returns A `QueryDirectory` object describing the contents of the directory, or `undefined` if
-   *   no queries were found.
+   * Return all known queries, represented as a tree.
+   *
+   * Trivial directories where there is only one child will be collapsed into a single node.
    */
-  private async discoverQueries(
-    workspaceFolders: readonly WorkspaceFolder[],
-  ): Promise<Array<FileTreeDirectory<string>>> {
-    const rootDirectories = [];
-    for (const workspaceFolder of workspaceFolders) {
-      const root = await this.discoverQueriesInWorkspace(workspaceFolder);
-      if (root !== undefined) {
-        rootDirectories.push(root);
+  public buildQueryTree(): Array<FileTreeNode<string>> {
+    const roots = [];
+    for (const workspaceFolder of getOnDiskWorkspaceFoldersObjects()) {
+      const queriesInRoot = this.paths.filter((query) =>
+        containsPath(workspaceFolder.uri.fsPath, query.path),
+      );
+      if (queriesInRoot.length > 0) {
```

Comment:
Could we invert this condition and do a `continue` if the `queriesInRoot.length === 0`? That would reduce the nesting here.